### PR TITLE
Add e2e for tranfer tokens with custom fees

### PR DESCRIPTION
**Pull request by IvanKavaldzhiev**
**Description**:

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by Nana-EC — COMMENTED** [review]
Nice.
Can we expose these in `itest` and `eet` also

**Comment by IvanKavaldzhiev** [comment]
> 

These tests are part of `CryptoTransferSuite` and `CryptoTransferHTSSuite`. They are both added in `itest` and `CryptoTransferSuite` is added in `eet`.

**Review by tinker-michaelj — APPROVED** [review]
Wow very clean solution and exhaustive testing! 🙏 Great work. 

Given your `BalanceChange.isForCustomFee()` flag, can we also fix our incorrect enforcement of `receiverSigRequired` on custom fee credits in the `TransferPrecompile` by changing [this line](https://github.com/hashgraph/hedera-services/blob/add-e2e-for-tranfer-tokens-with-custom-fees/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TransferPrecompile.java#L243) to
```
if (!change.isForCustomFee()) {
   ...
```
?

(The current check doesn't work because `numExplicitChanges` is [getting set](https://github.com/hashgraph/hedera-services/blob/add-e2e-for-tranfer-tokens-with-custom-fees/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/precompile/impl/TransferPrecompile.java#L209) **after** `extrapolateDetailsFromSyntheticTxn()`, so it actually includes the number of "implied" changes as well. 🤕)

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ledger/BalanceChange.java [hunk 1]
```diff
@@ -115,18 +117,24 @@ public static BalanceChange changingNftOwnership(
     }
 
     public static BalanceChange tokenAdjust(final Id account, final Id token, final long amount) {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ledger/BalanceChange.java:119]: IMO these `hbarAdjust` and `tokenAdjust` factories should be renamed to like `hbarCustomFeeAdjust()` to reduce the risk someone calls them outside a custom fee context.

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/CryptoTransferHTSSuite.java [hunk 2]
```diff
@@ -1538,6 +1568,1104 @@ private HapiApiSpec activeContractInFrameIsVerifiedWithoutNeedForSignature() {
...
+                        newKeyNamed(MULTI_KEY),
+                        newKeyNamed(RECEIVER_SIGNATURE),
+                        cryptoCreate(TOKEN_TREASURY),
+                        cryptoCreate(OWNER)
+                                .balance(100 * ONE_MILLION_HBARS)
+                                .maxAutomaticTokenAssociations(5)
+                                .key(MULTI_KEY),
+                        cryptoCreate(SENDER).balance(10 * ONE_HUNDRED_HBARS),
+                        cryptoCreate(RECEIVER)
+                                .balance(100 * ONE_MILLION_HBARS)
```
Comment (tinker-michaelj) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/CryptoTransferHTSSuite.java:1583]: There have been cases where our CI payer account has run out of money due to lots of high-balance accounts being created, so we might want to be less generous here. 😁

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ledger/BalanceChange.java (hunk 1 above)
Comment (IvanKavaldzhiev) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ledger/BalanceChange.java:119]: I was thinking the same thing 🤔 ✅

Comment (IvanKavaldzhiev) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ledger/BalanceChange.java]: DONE

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/CryptoTransferHTSSuite.java (hunk 2 above)
Comment (IvanKavaldzhiev) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/CryptoTransferHTSSuite.java]: Agree with you! :) DONE

**Review by stoqnkpL — APPROVED** [review]
LGTM

**Pull request merged**